core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/BaseAsyncClientHandler.java

```diff
@@ -77,16 +80,16 @@ protected BaseAsyncClientHandler(SdkClientConfiguration clientConfiguration,
 
         // For streaming requests, prepare() should be called as early as possible to avoid NPE in client
         // See https://github.com/aws/aws-sdk-java-v2/issues/1268
-        CompletableFuture<ReturnT> asyncTransformerFuture = asyncResponseTransformer.prepare();
+        AsyncStreamingResponseHandler<OutputT, ReturnT> asyncStreamingResponseHandler =
+            new AsyncStreamingResponseHandler<>(asyncResponseTransformer);
+        CompletableFuture<ReturnT> asyncTransformerFuture = asyncStreamingResponseHandler.prepare();
 
         ExecutionContext context = createExecutionContext(executionParams);
         context.executionAttributes().putAttribute(ASYNC_RESPONSE_TRANSFORMER_FUTURE, asyncTransformerFuture);
 
         HttpResponseHandler<OutputT> decoratedResponseHandlers =
             decorateResponseHandlers(executionParams.getResponseHandler(), context);
-
-        AsyncStreamingResponseHandler<OutputT, ReturnT> asyncStreamingResponseHandler =
-            new AsyncStreamingResponseHandler<>(asyncResponseTransformer, decoratedResponseHandlers);
+        asyncStreamingResponseHandler.responseHandler(decoratedResponseHandlers);
```

Comment:
Injecting a responseHandler after creation of asyncStreamingResponseHanlder is not ideal, but let me know if you have any other ideas. Basically, we need `asyncStreamingResponseHandler.prepare()` to be invoked upfront.

Reply:
Sounds like a challenge! It's not ideal at all...

Reply:
As discussed, we probably shouldn't capture the future in the `AsyncStreamingResponseHandler`, but I guess I'm okay with the mutation here. It should be on a volatile variable, though...

I still think this is more of an architectural issue than an unfixable problem.